### PR TITLE
Return enroll_status unspecified as default

### PR DESCRIPTION
**Pull request by flyinghermit**
Currently we return error when `enroll_status` is not supplied by the api client. This PR updates `enroll_status` to be `unspecified` as a default. 

**Backstory:**
When working with Terraform,  the `enroll_status` can be updated from both the terraform (`enrolled -> not_enrolled`) provider or from the teleport device trust service itself. When the `enroll_status` value changes from outside of Terraform (e.g., when enrolling device) resource config file, this can create issues for provider users to keep the data in sync with the Terraform state and resource config file(`.tf`).  

So for better user experience, terraform can exclude `enroll_status` in default state and only use the field to update the status `enrolled -> not_enrolled` when required.

### CODE REVIEW

**Review by r0mant — COMMENTED** [review]
Code looks ok but I'm not sure I fully understand how defaulting to "unspecified" really fixes the problem with the status changing outside of Terraform state from the PR description, couldn't it still happen regardless of what the initial device state is?

**Comment by sshahcodes** [comment]
Sorry for not being clear, this PR aims to improve terraform usage experience. You are correct that terraform will still have to sync state with teleport but defaulting will have better experience for admins as the "enroll_status" field in `.tf` files won't be a mandatory one. Admins can use that field only when they manually need to change state from `enrolled -> not_enrolled`, which will be pretty rare case.

File: api/types/device.go [hunk 1]
```diff
@@ -233,24 +227,20 @@ func ResourceEnrollStatusToString(enrollStatus devicepb.DeviceEnrollStatus) stri
 		return "enrolled"
 	case devicepb.DeviceEnrollStatus_DEVICE_ENROLL_STATUS_NOT_ENROLLED:
 		return "not_enrolled"
-	case devicepb.DeviceEnrollStatus_DEVICE_ENROLL_STATUS_UNSPECIFIED:
-		return "unspecified"
 	default:
-		return enrollStatus.String()
```
Comment (codingllama) [inline api/types/device.go:239]: This is mixing truly unknown proto values with "unspecified", a bad idea in terms of resistance to future changes. The changed code is not equivalent to what it was before.

File: api/types/device.go [hunk 2]
```diff
@@ -233,24 +227,20 @@ func ResourceEnrollStatusToString(enrollStatus devicepb.DeviceEnrollStatus) stri
...
 	switch enrollStatus {
 	case "enrolled":
-		return devicepb.DeviceEnrollStatus_DEVICE_ENROLL_STATUS_ENROLLED, nil
+		return devicepb.DeviceEnrollStatus_DEVICE_ENROLL_STATUS_ENROLLED
 	case "not_enrolled":
-		return devicepb.DeviceEnrollStatus_DEVICE_ENROLL_STATUS_NOT_ENROLLED, nil
-	case "unspecified":
-		return devicepb.DeviceEnrollStatus_DEVICE_ENROLL_STATUS_UNSPECIFIED, nil
+		return devicepb.DeviceEnrollStatus_DEVICE_ENROLL_STATUS_NOT_ENROLLED
 	default:
```
Comment (codingllama) [inline api/types/device.go:243]: Similarly to above, we are swallowing errors if we see a proto value we don't know about.

**Review by codingllama — CHANGES_REQUESTED** [review]
I'm not getting the point of the PR, as it stands it looks to me like a change for the worse.

Feel free to ping me for help.

File: api/types/device.go (hunk 2 above)
Comment (sshahcodes) [reply api/types/device.go:243]: What would be the consequence of swallowing error here? And otherwise, what purpose would "unspecified" address?

**Comment by codingllama** [comment]
I think we are going about Terraform the wrong way. There are some device fields that Terraform cannot, by definition, manage. For example, Terraform cannot unilaterally enroll a device, as this would break many of the guarantees of the enrollment system.

In light of that, I wonder if we should instead be more specific in what Terraform is capable of doing, and accept the fact that some of the device state evolves outside of it. For example, should Terraform hold the list of known devices, but ignore all other transient fields and server-managed state?

File: api/types/device.go (hunk 2 above)
Comment (sshahcodes) [reply api/types/device.go:243]: Yeah I assumed that. But the purpose of having "unspecified" state is still not clear.

**Comment by sshahcodes** [comment]
Actually that's the direction we are taking here. We will only support `asset_tag`, `os_type` and `enroll_status` for time being (at least until other requirement emerge). More details in the draft PR here - https://github.com/gravitational/teleport-plugins/pull/801. Among these three fields, only the value of `enroll_status` is updatable outside of `enrolldevice` process. And even for `enroll_status`, we only support updating status from `enrolled -> not_enrolled`. But this `enrolled -> not_enrolled` will be a rare use case and making it mandatory field will only make configuration complicated for terraform maintainers. 

> accept the fact that some of the device state evolves outside of it. 

This is considered in the current implementation and that's the reason for this PR to make terraform maintainer's job easy by letting them ignore field (in this case `enroll_status`) be an optional in terraform. Otherwise they will have hard time syncing terraform state with hcl file. 

> For example, should Terraform hold the list of known devices, but ignore all other transient fields and server-managed state?

This is already supported by terraform. But since we return error when the transient fields are not properly defined, this functionality breaks. Hence user will either have to keep track of (and update) these transient field or we can make these values optional.

File: api/types/device.go (hunk 2 above)
Comment (codingllama) [reply api/types/device.go:243]: Do you mean in general? Calling the zero value enum value "UNSPECIFIED" is a proto best practice: https://protobuf.dev/programming-guides/style/#enums. Ping me on chat if you want further explanation.

**Comment by codingllama** [comment]
> This is already supported by terraform. But since we return error when the transient fields are not properly defined, this functionality breaks. Hence user will either have to keep track of (and update) these transient field or we can make these values optional.

That sounds wrong, the server doesn't require the enroll_status field to be set. I suspect you may be having problems specifically while converting the empty string (`""`) to an enroll status enum; if that's the case, we can handle just that, but please leave other parts of the functions unchanged.

**Comment by sshahcodes** [comment]
> the server doesn't require the enroll_status field to be set

Correct. It's the API client we wrote for terraform expects that - https://github.com/gravitational/teleport/blob/master/api/types/device.go#L85 (which this PR tries to update). 

Noted on just handling the empty or null type and including those case for "unspecified". Though I am not sure the reason we support `unspecified`.

**Review by codingllama — APPROVED** [review]

**Review by r0mant — APPROVED** [review]

**Pull request merged**